content/influxdb/cloud-dedicated/reference/influxql/internals.md

```diff
@@ -0,0 +1,163 @@
+   points. The emitter's job is to convert simple time/value points into the
+   more complex result objects that are returned to the client.
+
+### Understanding iterators
+
+Iterators  provide a simple interface for looping over a set of points.
+For example, this is an iterator over Float points:
+
+```
+type FloatIterator interface {
+    Next() *FloatPoint
+}
+```
+
+These iterators are created through the `IteratorCreator` interface:
+
+```
+type IteratorCreator interface {
+    CreateIterator(opt *IteratorOptions) (Iterator, error)
+}
+```
+
+The `IteratorOptions` provide arguments about field selection, time ranges,
+and dimensions that the iterator creator can use when planning an iterator.
+The `IteratorCreator` interface is used at many levels such as the `Shards`,
+`Shard`, and `Engine`. This allows optimizations to be performed when applicable
+such as returning a precomputed `COUNT()`.
+
+Iterators aren't just for reading raw data from storage though. Iterators can be
+composed so that they provided additional functionality around an input
+iterator. For example, a `DistinctIterator` can compute the distinct values for
+each time window for an input iterator. Or a `FillIterator` can generate
+additional points that are missing from an input iterator.
+
+This composition also lends itself well to aggregation. For example, a statement
+such as this:
+
+```sql
+SELECT MEAN(value) FROM cpu GROUP BY time(10m)
+```
+
+In this case, `MEAN(value)` is a `MeanIterator` wrapping an iterator from the
+underlying shards. However, if we can add an additional iterator to determine
+the derivative of the mean:
+
+```sql
+SELECT DERIVATIVE(MEAN(value), 20m) FROM cpu GROUP BY time(10m)
+```
+
+### Cursors
+
+A **cursor** identifies data by shard in tuples (time, value) for a single series (measurement, tag set and field). The cursor trasverses data stored as a log-structured merge-tree and handles deduplication across levels, tombstones for deleted data, and merging the cache (Write Ahead Log). A cursor sorts the `(time, value)` tuples by time in ascending or descending order.
+
+For example, a query that evaluates one field for 1,000 series over 3 shards constructs a minimum of 3,000 cursors (1,000 per shard).
+
+### Auxiliary fields
+
+Because InfluxQL allows users to use selector functions such as `FIRST()`,
+`LAST()`, `MIN()`, and `MAX()`, the engine must provide a way to return related
+data at the same time with the selected point.
+
+Let's look at the following query:
+
+```sql
+SELECT FIRST(value), host FROM cpu GROUP BY time(1h)
+```
+
+We are selecting the first `value` that occurs every hour but we also want to
+retrieve the `host` associated with that point. Since the `Point` types only
+specify a single typed `Value` for efficiency, we push the `host` into the
+auxiliary fields of the point. These auxiliary fields are attached to the point
+until it is passed to the emitter where the fields get split off to their own
+iterator.
```

Comment:
Maybe not too important, but I think this could be clearer - especially the intro sentence. It's not clear immediately why selector functions force the need to return related data.... Could probably just state upfront what we mean by "auxiliary fields" for a `Point`.